src/main/java/htsjdk/io/AsyncWriterPool.java

```diff
@@ -0,0 +1,252 @@
+     *
+     * @param threads max number of threads to use
+     */
+    public AsyncWriterPool(final int threads) {
+        if (threads < 1) throw new IllegalArgumentException("Threads must be >= 1: " + threads);
+
+        this.executor = Executors.newWorkStealingPool(threads);
+    }
+
+    /**
+     * Create an AsyncWriterPool using all available processors. The number of threads in use at one time will grow and
+     * shrink with load.
+     */
+    public AsyncWriterPool() {
+        this(Runtime.getRuntime().availableProcessors());
+    }
+
+    /**
+     * Asynchronously closes each writer in the pool. Each writer calls {@code .close()} in a CompletableFuture, which
+     * allows the writers to wait for any ongoing writes, drain any remaining elements from the queue, and then close
+     * the inner writer. All writers will immediately cease to accept new items to write and any call to {@code .write()}
+     * after calling this method will throw an exception. This method waits for all {@code CompletableFuture}s to
+     * complete and then shuts down the executor.
+     * <p>
+     * This method blocks and till all writers have closed and pool has shut down.
+     *
+     * @throws IOException if any writer raises an exception
+     */
+    @Override
+    public void close() throws IOException {
+        if (this.poolClosed) return;
+        this.poolClosed = true;
+        CompletableFuture.allOf(this.writers.stream().map(PooledWriter::nonBlockingClose).toArray(CompletableFuture[]::new)).join();
+        this.executor.shutdown();
+    }
+
+    /**
+     * Exchange a class implementing {@link Writer} for a {@link PooledWriter}.
+     *
+     * @param writer         a class implementing {@link Writer}
+     * @param queue          a queue to use for this writer, bounded or unbounded.
+     * @param writeThreshold the minimum number of items needed before scheduling a thread for writing.
+     * @param <A>            the type of the items the {@link Writer} can write.
+     * @return a {@link PooledWriter}
+     */
+    public <A> Writer<A> pool(final Writer<A> writer, final BlockingQueue<A> queue, final int writeThreshold) {
+        PooledWriter<A> pooledWriter = new PooledWriter<>(writer, queue, writeThreshold);
+        this.writers.add(pooledWriter);
+        return pooledWriter;
+    }
+
+
+    /**
+     * Any class that implements {@link Writer} can be exchanged for a {@code PooledWriter}. The PooledWriter provides
+     * the same API as {@link Writer}, but will manage buffering of writes and sending to the {@link AsyncWriterPool} it
+     * is attached to.
+     * <p>
+     * Note, {@code PooledWriter} is not intrinsically threadsafe, and any writing / closing done across multiple
+     * threads should be synchronized.
+     *
+     * @param <A> the type of the item the writer can write
+     */
+    private class PooledWriter<A> implements Writer<A> {
```

Comment:
My instinct is exactly the opposite of Tim's here.  I was going to say make this static.

Reply:
I think this has to be non-static so that it can access the pool instance of the outer class. I'm not a java expert though and defer to either your @tfenne!

Reply:
It either has to be non-static, _or_ take in a reference to the pool in it's constructor.  But yes, one way or another it needs a reference to the pool instance.  I don't have strong feelings either way on whether that's by being a non-static inner class, or by taking the pool directly in the constructor.

Reply:
Yeah, my feeling is always static if you can and just pass in through the constructor, but it's not a big deal here either way.  I just thought it was funny that we basically had opposite feelings about the cleanest way to do it.